src/utils/ohlc_fetcher.py

```diff
@@ -0,0 +1,149 @@
+import calendar
+import datetime as dt
+import json
+import time
+
+import pandas as pd
+import requests
+from sqlalchemy import create_engine, desc, select
+
+engine = create_engine(
+    "postgresql+psycopg2://Jayc:admin@127.0.0.1:5432/bigaluto_stockdb"
+)
+
+from sqlalchemy.orm import Session
+
+from ..sql.schema import Data, Ticker, Timeframe
+
+
+def execute_fetcher(
+    ticker_collection: dict[str, list[tuple[int, str]]],
+):
+    headers = {
+        "Accept": "*/*",
+        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36",
+    }
+
+    all_errors = []
+    timeframe_win = "D"
+
+    for date, ticker_list in ticker_collection.items():
+        start_date = dt.datetime.strptime(date, "%Y-%m-%d").date()
+        for ticker_id, _ in ticker_list:
+            try:
+                unixts_1 = int(calendar.timegm(start_date.timetuple()))
+                unixts_2 = int(calendar.timegm(dt.datetime.utcnow().timetuple()))
+
+                query = f"https://query2.finance.yahoo.com/v8/finance/chart/NVR?formatted=true&crumb=0VkTr5j7Fso&lang=en-GB&region=US&includeAdjustedClose=true&interval=1d&period1={unixts_1}&period2={unixts_2}&events=capitalGain%7Cdiv%7Csplit&useYfid=true&corsDomain=uk.finance.yahoo.com"
+                r = requests.get(query, headers=headers)
+                content = r.json()
+
+                content_data = content["chart"]["result"][0]
+
+                if "timestamp" not in content_data:
+                    continue
+
+                timestamp_list = content_data["timestamp"]
+                events_content = content_data["indicators"]
+                ohlc_content = events_content["quote"][0]
+                adj_close = events_content["adjclose"][0]
+
+                df = pd.DataFrame(
+                    {
+                        "date": timestamp_list,
+                        "open": ohlc_content["open"],
+                        "high": ohlc_content["high"],
+                        "low": ohlc_content["low"],
+                        "close": ohlc_content["close"],
+                        "adj_close": adj_close["adjclose"],
+                        "volume": ohlc_content["volume"],
+                    }
+                )
+                df["date"] = pd.to_datetime(df["date"], unit="s").dt.date
+
+                with Session(engine) as session:
```

Comment:
make the DB operations more generic and put into util file, so we can simply use db_util.create_enitites() / db_util.delete_entities() etc. to interact with DB in the functions